### PR TITLE
Use MarkDistinct only in limited cases

### PR DESCRIPTION
**Pull request by lukasz-stec**
## Description

MarkDistinct is beneficial only when there is
a limited parallelism in query (e.g. global distinct aggregation is handled by a single thread).
For cases where there is enough parallelism, MarkDistinct hurts performance due to excessive data shuffle
and introducing multiple stages in case of
multiple distinct aggregations.

This PR also changes the type of `use_mark_distinct` session property and `optimizer.use-mark-distinct` config property from `Boolean` to `Varchar`. The new property still accepts `true` and `false` values to keep backwards compatibility but those have to be specified as Varchars (e.g. single quotes in the cli)

## Additional context and related issues
To test different possible approaches I used queries that use different count of distinct values in a group by and 1 or 2 distinct aggregations + one non distinct aggregation.
I used three different cardinalities:
small: 502 unique values
medium: 10K unique values
high 240M  unique values

The tests show that for small NDV mark distinct works better and for high NDV distinct  aggregation is better.
In the medium case, `optimize_mixed_distinct_aggregations` is better for the duration at a significant CPU cost.

```
Queries
small cardinality, one distinct aggr

SELECT ss_store_sk, Sum(ss_quantity), Count(DISTINCT ss_customer_sk) FROM hive.tpcds_sf1000_orc.store_sales GROUP BY ss_store_sk;

small cardinality, 2 distinct aggrs

SELECT ss_store_sk, Sum(ss_quantity), Count(DISTINCT ss_customer_sk), Count(DISTINCT ss_promo_sk) FROM hive.tpcds_sf1000_orc.store_sales GROUP BY ss_store_sk;

 medium cardinality, 1 distinct aggr

SELECT ss_ticket_number % 10000, Sum(ss_quantity), Count(DISTINCT ss_customer_sk) FROM hive.tpcds_sf1000_orc.store_sales GROUP BY 1;

medium cardinality, 2 distinct aggrs

SELECT ss_ticket_number % 10000, Sum(ss_quantity), Count(DISTINCT ss_customer_sk), Count(DISTINCT ss_promo_sk) FROM hive.tpcds_sf1000_orc.store_sales GROUP BY 1;

big cardinality, 1 distinct aggr

SELECT ss_ticket_number, Sum(ss_quantity), Count(DISTINCT ss_customer_sk) FROM hive.tpcds_sf1000_orc.store_sales GROUP BY ss_ticket_number;

big cardinality, 2 distinct aggrs

SELECT ss_ticket_number, Sum(ss_quantity), Count(DISTINCT ss_customer_sk), Count(DISTINCT ss_promo_sk) FROM hive.tpcds_sf1000_orc.store_sales GROUP BY ss_ticket_number;
```
| partitioning | aggregation mode                     | cardinality | \# distinct aggregations | duration | CPU  |
| ------------ | ------------------------------------ | ----------- | ------------------------ | -------- | ---- |
| unpart       | optimize_mixed_distinct_aggregations | small       | 1                        | 10.41    | 1661 |
| unpart       | MarkDistinct                         | small       | 1                        | 13.27    | 1176 |
| unpart       | distinct aggregation                 | small       | 1                        | 53.58    | 1193 |
| unpart       | MarkDistinct                         | small       | 2                        | 32.37    | 2813 |
| unpart       | distinct aggregation                 | small       | 2                        | 1:15     | 1902 |
| unpart       | optimize_mixed_distinct_aggregations | large       | 1                        | 19.77    | 2343 |
| unpart       | MarkDistinct                         | large       | 1                        | 20.52    | 1446 |
| unpart       | distinct aggregation                 | large       | 1                        | 18.32    | 1315 |
| unpart       | MarkDistinct                         | large       | 2                        | 1:07     | 6591 |
| unpart       | distinct aggregation                 | large       | 2                        | 35.63    | 3609 |
| unpart       | optimize_mixed_distinct_aggregations | medium      | 1                        | 11.14    | 1637 |
| unpart       | MarkDistinct                         | medium      | 1                        | 14.66    | 1219 |
| unpart       | distinct aggregation                 | medium      | 1                        | 14.06    | 1190 |
| unpart       | MarkDistinct                         | medium      | 2                        | 36.71    | 3641 |
| unpart       | distinct aggregation                 | medium      | 2                        | 20.62    | 2446 |
| part         | optimize_mixed_distinct_aggregations | small       | 1                        | 11.57    | 1721 |
| part         | MarkDistinct                         | small       | 1                        | 15.12    | 1175 |
| part         | distinct aggregation                 | small       | 1                        | 55.74    | 1188 |
| part         | MarkDistinct                         | small       | 2                        | 34.6     | 2837 |
| part         | distinct aggregation                 | small       | 2                        | 1:14     | 1863 |
| part         | optimize_mixed_distinct_aggregations | large       | 1                        | 24.1     | 2619 |
| part         | MarkDistinct                         | large       | 1                        | 24.91    | 1561 |
| part         | distinct aggregation                 | large       | 1                        | 18.77    | 1349 |
| part         | MarkDistinct                         | large       | 2                        | 1:08     | 6701 |
| part         | distinct aggregation                 | large       | 2                        | 35.27    | 3623 |
| part         | optimize_mixed_distinct_aggregations | medium      | 1                        | 13.07    | 1820 |
| part         | MarkDistinct                         | medium      | 1                        | 14.81    | 1263 |
| part         | distinct aggregation                 | medium      | 1                        | 14.15    | 1216 |
| part         | MarkDistinct                         | medium      | 2                        | 36.5     | 3574 |
| part         | distinct aggregation                 | medium      | 2                        | 19.67    | 2461 |

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(X) Release notes are required, with the following suggested text:

```markdown
# General
* Improve performance of queries with distinct aggregations using table statistics.
The choice of whether `MarkDistinct` operator is used for distinct aggregations is now configurable using a new configuration property `optimizer.mark-distinct-strategy` with the values `NONE`, `ALWAYS` and `AUTOMATIC`, with
`AUTOMATIC` as the default.
The `optimizer.use-mark-distinct` configuration property is deprecated in favour of the new property. If used, it's values are mapped to the `optimizer.mark-distinct-strategy` property as `true` -> `AUTOMATIC`, `false` -> `NONE`.
```

### CODE REVIEW

**Review by Dith3r — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 1]
```diff
@@ -531,11 +534,15 @@ public SystemSessionProperties(
...
+                new PropertyMetadata<>(
                         USE_MARK_DISTINCT,
                         "Implement DISTINCT aggregations using MarkDistinct",
-                        optimizerConfig.isUseMarkDistinct(),
-                        false),
+                        createUnboundedVarcharType(),
+                        UseMarkDistinct.class,
+                        optimizerConfig.getUseMarkDistinct(),
+                        false,
+                        value -> UseMarkDistinct.fromPropertyValue((String) value),
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:544]: Why is this needed? See how `JOIN_DISTRIBUTION_TYPE` is handled

File: core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java [hunk 2]
```diff
@@ -471,15 +473,15 @@ public OptimizerConfig setOptimizeMetadataQueries(boolean optimizeMetadataQuerie
...
 
-    public boolean isUseMarkDistinct()
+    public UseMarkDistinct getUseMarkDistinct()
     {
         return useMarkDistinct;
     }
 
     @Config("optimizer.use-mark-distinct")
-    public OptimizerConfig setUseMarkDistinct(boolean value)
+    public OptimizerConfig setUseMarkDistinct(String value)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:482]: This is backward incompatible change.

We should add another property:

```
optimizer.mark-distinct-strategy=NONE|ALWAYS|AUTOMATIC
```
(consistent with `optimizer.join-reordering-strategy`). 

Then we can map `optimizer.use-mark-distinct=true` -> `optimizer.mark-distinct-strategy=AUTOMATIC`
and `optimizer.use-mark-distinct=true` -> `optimizer.mark-distinct-strategy=NONE`.
We can also mark `optimizer.use-mark-distinct` as deprecated.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java [hunk 3]
```diff
@@ -165,4 +203,45 @@ public Result apply(AggregationNode parent, Captures captures, Context context)
...
+        if (aggregationNode.getGroupingKeys().isEmpty()) {
+            // global distinct aggregation is computed using a single thread. MarkDistinct will help parallelize the execution.
+            return true;
+        }
+        double numberOfDistinctValues = context.getStatsProvider().getStats(aggregationNode).getOutputRowCount();
+        if (Double.isNaN(numberOfDistinctValues)) {
+            // if the estimate is unknown, use MarkDistinct to avoid query failure
+            return true;
+        }
+        int numberOfThreadsInACluster = taskCountEstimator.estimateSourceDistributedTaskCount(context.getSession()) * getTaskConcurrency(context.getSession());
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:218]: mark distinct is computed using `estimateHashedTaskCount` task count

Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:218]: it's not number of threads, but number of tasks

File: core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java [hunk 4]
```diff
@@ -648,28 +651,28 @@ public void testTopNRowNumberIsExactlyPartitioned()
...
-                "        VALUES\n" +
-                "            (1, 2)\n" +
-                "    ) t (a, b)\n" +
-                "    GROUP BY\n" +
-                "        a,\n" +
-                "        b\n" +
-                ")\n" +
-                "LIMIT\n" +
-                "    2",
+                        "    a,\n" +
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestAddExchangesPlans.java:654]: formatting, separate commit

**Review by sopel39 — COMMENTED** [review]
Some comments. I think the approach is fine (although it might hurt from NDV misestimation, but it's life). Please run benchmarks. I'm curious if we will see some gains

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java [hunk 5]
```diff
@@ -165,4 +203,45 @@ public Result apply(AggregationNode parent, Captures captures, Context context)
...
+
+        if (numberOfDistinctValues <= MARK_DISTINCT_MAX_OUTPUT_ROW_COUNT_MULTIPLIER * numberOfThreadsInACluster) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:220]: One disadvantage of not using mark distinct is that without mark distinct partial aggregations are not possible. Maybe it's not a problem, but I could imagine it might be if data is already pre-partitioned as mark distinct requires

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java [hunk 6]
```diff
@@ -165,4 +203,45 @@ public Result apply(AggregationNode parent, Captures captures, Context context)
...
+            // small numberOfDistinctValues reduces the distinct aggregation parallelism, also because the partitioning may be skewed.
+            // MarkDistinct should increase the parallelism at a cost of CPU.
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:222]: It's not only about parallelism, but also memory usage (update comment) which would concentrate on few nodes.

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestMultipleDistinctAggregationToMarkDistinct.java [hunk 7]
```diff
@@ -14,16 +14,38 @@
...
+import static io.trino.sql.planner.assertions.PlanMatchPattern.aggregation;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.functionCall;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.globalAggregation;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.markDistinct;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.singleGroupingSet;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.values;
 import static io.trino.sql.planner.iterative.rule.test.PlanBuilder.expression;
+import static io.trino.sql.planner.plan.AggregationNode.Step.SINGLE;
 
 public class TestMultipleDistinctAggregationToMarkDistinct
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestMultipleDistinctAggregationToMarkDistinct.java:43]: nit: We could also disable `MarkDistinct` when number of stages would exceed limit, but this can be another PR and it's not that simple...

File: core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java (hunk 2 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:482]: This brings some problems because now we have two properties controlling single behavior. What if someone sets `use_mark_distinct= false` and `optimizer.mark-distinct-strategy=ALWAYS`?
Should we force users to match? what if user wants optimizer.mark-distinct-strategy=ALWAYS?

**Comment by lukasz-stec** [comment]
tpch/tpcds orc part sf1k benchamrks look good (4 to 5 % CPU + similar fo duration) but this may be variability as queries that don't have distinct aggregation also speed up (e.g. tpch q17, 18, q21 ).

image

[oss-orc-part-sf1K-020223.pdf](https://github.com/trinodb/trino/files/10580452/oss-orc-part-sf1K-020223.pdf)

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java (hunk 1 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:544]: > this was to support old values (true and false). Will drop it in place of new property

I don't think it does. `set session x=true` is different than `set session x='true'`

File: core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java (hunk 2 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:482]: You can make the old one `nullabe` to check if it was set explicitly and prevent invalid config

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java [hunk 8]
```diff
@@ -62,6 +69,25 @@
...
+    {
+        AUTO, FORCE, NO;
+
+        public static UseMarkDistinct fromPropertyValue(String value)
+        {
+            value = value.toLowerCase(ENGLISH).trim();
+            return switch (value) {
+                case "true", "auto" -> AUTO;
+                case "force" -> FORCE;
+                case "false", "no" -> NO;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:82]: "no" -> "NEVER"

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java [hunk 9]
```diff
@@ -165,4 +203,46 @@ public Result apply(AggregationNode parent, Captures captures, Context context)
...
                         .build());
     }
+
+    private boolean shouldAddMarkDistinct(AggregationNode aggregationNode, Context context)
+    {
+        if (aggregationNode.getGroupingKeys().isEmpty()) {
+            // global distinct aggregation is computed using a single thread. MarkDistinct will help parallelize the execution.
+            return true;
+        }
+        double numberOfDistinctValues = context.getStatsProvider().getStats(aggregationNode).getOutputRowCount();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:213]: Are we sure that relying on aggregation output row count is safe ?
For multiple grouping keys, this is likely to be an overestimate (assumption that there is no correlation between columns). For non-partitioned columns of partitioned tables in hive connector we are underestimating NDVs by taking a max of NDVs from all partitions. Also, we sample statistics from only 100 partitions.
All these problems exist in existing usages as well (ApplyPreferredTableWriterPartitioning in particular). However, it seems risky to increase dependence on this even after we know of cases where bad statistics for the above reasons have caused problems in existing usages.

**Comment by sopel39** [comment]
> tpch/tpcds orc part sf1k benchamrks look good (4 to 5 % CPU + similar fo duration) but this may be variability as queries that don't have distinct aggregation also speed up (e.g. tpch q17, 18, q21 ).

The network didn't go down so I'm not sure that rule triggered

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java (hunk 9 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:213]: > Are we sure that relying on aggregation output row count is safe ?

I'm sure there are cases (over-estimation of very small NDV) where relying on stats here will cause the rule to choose the wrong approach, but that goes for most CBO rules, right?
I can also increase the thresholds/multipliers to count overestimation in but that has negative consequences if the estimation is correct.

**Comment by lukasz-stec** [comment]
> > tpch/tpcds orc part sf1k benchamrks look good (4 to 5 % CPU + similar fo duration) but this may be variability as queries that don't have distinct aggregation also speed up (e.g. tpch q17, 18, q21 ).
>
> The network didn't go down so I'm not sure that rule triggered

@sopel39 I checked, there are no `tpch/tpcds` queries that would want to use distinct aggregation. The rule triggers for some queries that have global aggregation but uses `MarkDistinct` for those as expected.

**Comment by lukasz-stec** [comment]
Following the discussion, I changed the approach to `use_mark_distinct` property. Instead of changing its type to varchar, which makes it backward incompatible, I added a new property `mark_distinct_strategy` that overrides the `use_mark_distinct` value if both are set.
Another open issue, raised by @raunaqmorarka here https://github.com/trinodb/trino/pull/15927#discussion_r1104145620, is: can we rely on NDV stats to disable `MarkDistinct`? There is a tradeoff here that in most cases this should work fine or at least not make it worse, except when we overestimate small NDV (e.g. we estimate tens of thousands instead of tens of distinct values). This can happen with multiple, strongly correlated grouping keys where one has low cardinality.
cc @sopel39 @raunaqmorarka

File: core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java (hunk 2 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java]: added mark-distinct-strategy

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java (hunk 9 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:202]: I restricted reliance on the NDV stats only for a single column group by to avoid issues with overestimation

**Review by raunaqmorarka — COMMENTED** [review]
lgtm

File: core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java [hunk 10]
```diff
@@ -471,18 +482,36 @@ public OptimizerConfig setOptimizeMetadataQueries(boolean optimizeMetadataQuerie
...
+    @Nullable
+    public Boolean isUseMarkDistinct()
     {
         return useMarkDistinct;
     }
 
+    @Deprecated
     @Config("optimizer.use-mark-distinct")
-    public OptimizerConfig setUseMarkDistinct(boolean value)
+    @LegacyConfig(value = "optimizer.use-mark-distinct", replacedBy = "optimizer.mark-distinct-strategy")
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:494]: @raunaqmorarka `@LegacyConfig` causes:
```
Error:  io.trino.TestJdbcResultSetCompatibilityOldDriver.setupServer  Time elapsed: 1.726 s  <<< FAILURE!
io.airlift.bootstrap.ApplicationConfigurationException: 
Configuration errors:

1) Error: @Config method [public OptimizerConfig OptimizerConfig.setUseMarkDistinct(Boolean)] has annotation claiming to be replaced by another property ('optimizer.mark-distinct-strategy')

2) Error: @Config property name 'optimizer.use-mark-distinct' appears in @LegacyConfig annotation for method [public OptimizerConfig OptimizerConfig.setUseMarkDistinct(Boolean)]

```
https://github.com/trinodb/trino/actions/runs/4292550494/jobs/7479120211

Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:493]: This is because  `@LegacyConfig` does not mix with `@Config`. Fixed

**Review by raunaqmorarka — APPROVED** [review]
